include/sys/spa.h

```diff
@@ -408,6 +488,26 @@ _NOTE(CONSTCOND) } while (0)
 #define	BP_GET_LEVEL(bp)		BF64_GET((bp)->blk_prop, 56, 5)
 #define	BP_SET_LEVEL(bp, x)		BF64_SET((bp)->blk_prop, 56, 5, x)
 
+/* encrypted, authenticated, and MAC cksum bps use the same bit */
+#define	BP_USES_CRYPT(bp)		BF64_GET((bp)->blk_prop, 61, 1)
+#define	BP_SET_CRYPT(bp, x)		BF64_SET((bp)->blk_prop, 61, 1, x)
+
+#define	BP_IS_ENCRYPTED(bp)			\
+	(BP_USES_CRYPT(bp) &&			\
+	BP_GET_LEVEL(bp) <= 0 &&		\
+	DMU_OT_IS_ENCRYPTED(BP_GET_TYPE(bp)))
```

Comment:
My biggest concern there is that we would be losing redundancy unnecessarily. Currently, if you set copies=2, ZFS will give you 3 copies of your metadata. With the current implementation, any data that is authenticated can still have all 3 copies.

I'm also not completely convinced that the structure of the bp should not be based on the `dmu_ot` table. After all `DMU_OT_INTENT_LOG` blocks don't really use the checksum field in the conventional way and `blk_fill` has a different meaning for `DMU_OT_DNODE` blocks than for every other block type. The only reason these examples haven't caused a problem and the encryption changes have is because we don't have special printing for them.